### PR TITLE
Add exec block to the instance resource

### LINKED ISSUE #54: how to perform "remote-exec", but inside the LXD
Terraform version: Terraform v0.9.6
terraform-provider-lxd: 0.9.5 beta1

As part of the LXD creation, I push some config files and a service startup script (upstart)
These cannot be baked into the LXD image.

It would be quite useful if I could target command execution directly inside the LXD container.

For now, I am making do with a "remote-exec" on the LXD server and do:
```lxd exec mylxd1 -- telinit 2```  (to make the upstart script start)

Regards,
Shantanu

### PR DESCRIPTION
**Pull request by MusicDin**
This PR adds `exec` block to the instance resource which allows running commands within the instance.

- [x] Requires #385 to be rebased.
- [x] Requires #387 to be able to properly manage `exec` state.

Fixes #54

### CODE REVIEW

**Comment by MusicDin** [comment]
@simondeziel Your comment from #387 PR:

> One last thing regarding the "working devlxd communication" aspect. How does that degrades if someone has security.devlxd=false? For containers, this is perfectly usable but I could see someone using it on VMs too where to connection mean is either the lxc console or ssh.

Just to not forget about it.

**Comment by simondeziel** [comment]
> @simondeziel Your comment from #387 PR:
> 
> > One last thing regarding the "working devlxd communication" aspect. How does that degrades if someone has security.devlxd=false? For containers, this is perfectly usable but I could see someone using it on VMs too where to connection mean is either the lxc console or ssh.
> 
> Just to not forget about it.

To clarify, I was wrong and `security.devlxd=false` does not affect the ability to `exec` into VMs.

**Review by simondeziel — COMMENTED** [review]
I like the proposal but have a few comments (as always ;)

I think Puppet really provides an easy to work with `exec` resource that we should take inspiration from.

File: internal/instance/resource_instance_test.go
Comment (MusicDin) [reply internal/instance/resource_instance_test.go:1317]: This is supposed to test whether special characters such as backslash (`\`) and pipe (`|`) are correctly passed to the instance. But yeah, could be for sure something something more readable :sweat_smile:

File: internal/instance/resource_instance.go
Comment (MusicDin) [reply internal/instance/resource_instance.go:327]: Good idea. Will add, thanks.

File: docs/resources/instance.md
Comment (MusicDin) [reply docs/resources/instance.md:149]: To recap our yesterday's conversation. The tricky part about special conditions is that we need to evaluate the plan of an instance before the update without doing any modification on the infrastructure itself. This means that any checks that need to be executed within an instance, would require instance to be running, which we *can not* ensure during the planning phase. For VMs, even simple file require the instance to be running.

Technically, we could ensure the instance is running during the planning phase by simply starting it if it is stopped, but this goes against Terraform concepts because we are doing an *update* without prior approval from the user. In addition, planning phase could take up to a minute instead of a few seconds.

Another approach would be to treat those special conditions as *requires update*, which means that such instances would always show *plan changes*, even if the conditions are already satisfied. Conditions would then be evaluated only during an actual update.

However, running a command on each apply can be already achieved using triggers:
```hcl
exec {
  command  = ["..."]
  triggers = [timestamp()] # This will ensure the change is always detected.
}
```

With such approach we can add new conditions that will determine whether command needs to be run, but they should not trigger any plan changes, because we cannot verify whether the change is already satisfied.

For advance use cases, the easiest approach would be to upload Puppet/Ansible configuratio to an instance (using `file` block) and then apply it using `exec` block that has a `timestamp()` trigger.

File: internal/common/lxd_exec.go
Comment (adamcstephens) [inline internal/common/lxd_exec.go:44]: I don't like hardcoding the interpreter to `/bin/sh`. It assumes that this path will exist, which is not necessarily true, and can be a different package depending on the image being worked with. Could we instead allow a user to provide their own interpreter rather than selecting a generic `shell` value?

Comment (MusicDin) [reply internal/common/lxd_exec.go:44]: Yeah, by default we do not use any interpreter, so the user can provide the command as:
```hcl
command = ["/bin/sh", "-c", "..."]
```

I though about having an interpreter as a list of strings (terraform's [local-exec](https://developer.hashicorp.com/terraform/language/resources/provisioners/local-exec#interpreter) uses that approach):
```hcl
command     = ["..."]
interpreter = ["/bin/sh", "-c"]
```

But it seems like just splitting a list into two lists, or? 

Could we maybe have `interpreter` attribute as you have proposed, but have `command` as a string? - However, how would then a simple command like `apt update` look like if we want to avoid using shell?
```hcl
command     = "apt update"
interpreter = ["sh", "-c"]

command     = "update"
interpreter = ["apt"]
```

Maybe the interpreter *should* be the required field?

Comment (simondeziel) [reply internal/common/lxd_exec.go:44]: If you take that route, might as well use this form:
```
command   = "apt" # note the bare string, not an array/list
arguments = ["dist-upgrade", "-V", "-y"]
```

Comment (adamcstephens) [reply internal/common/lxd_exec.go:44]: Some prior art would be the built-in local-exec provisioner. It sets a default of `/bin/sh -c ` (in a slice) if there is not a specified parameter and takes a simple string as the command. `working_dir` and `environment` may not be bad to offer either.

https://developer.hashicorp.com/terraform/language/resources/provisioners/local-exec
https://github.com/hashicorp/terraform/blob/main/internal/builtin/provisioners/local-exec/resource_provisioner.go

I'd be fine with defaulting to `/bin/sh` if there is an alternative

Comment (MusicDin) [reply internal/common/lxd_exec.go:44]: How about we have three attributes? - `posix`, `command`, `interpreter`

posix would be mutually exclusive with command and interpreter.
*I don't like posix as attribute name, but just for demo*

So user can provide either:
`posix = ["ls", "/"]` // So that we allow executing commands without shell

or:
```hcl
command     = "ls /"
interpreter = ["bash", "-c"] # Defaults to /bin/sh -c
```

With this approach, we could maybe also have list of `commands` instead of single command per exec block?

Comment (MusicDin) [reply internal/common/lxd_exec.go:44]: @terraform-lxd/maintainers Any comments/suggestions on https://github.com/terraform-lxd/terraform-provider-lxd/pull/386#discussion_r1421009176?

Comment (adamcstephens) [reply internal/common/lxd_exec.go:44]: I think my preference would be just using `command` and `interpreter`, with `interpreter` defaulting to `bash -c` and `command` being a string. This should support simple cases of just wanting to run a command and not worrying about the interpreter, but also allow for more complex cases where specifying the interpreter is desirable. I'm not sure if I see the value in also having a `posix` attribute, unless I'm missing something.

Comment (MusicDin) [reply internal/common/lxd_exec.go:44]: Thanks everyone. 

Based on the comments, I will leave `command` as is (list of strings), and change the `interpreter` attribute from string (enum) to list of strings. If interpreter is set, it will be simply prepended to the command.

**Review by simondeziel — APPROVED** [review]

**Pull request merged**
